### PR TITLE
[Admin] Improve tree view perfomance for non-admin users

### LINKED ISSUE #6383: [Core][DataObjects] Severe performance degradation when opening folders with many objects in the tree
# Bug Report

Hello,
we have a project with 19 `DataObjects` definitions and a large number of objects (nearly 1 million in total). The majority of the objects are of one type and located in one folder. 
If a user with limited permissions (no admin rights) opens this folder, loading the tree or grid will take a very, very long time (we're talking here about minutes...) The request blocks other requests and possibly the whole system, which makes it impossible to work with the system during this time.

### Expected behaviour
Context: user with limited permissions

A tree-list or grid-view of folders with a large number of objects are loaded in seconds, not minutes (as it is with an admin user).

### Actual behaviour
If you click on a folder with thousands of thousands of items, it takes minutes to load and display the data. (If the request is not already timed out anyway)

### Steps to reproduce
1. create a folder
2. create a large number of objects (600.000 or more) with the folder as the parent
3. create an user with limited permissions. give user permission to "list" the content of the folder 
4. log in with the limited user in another browser
5. open tree or grid for created folder.
6. wait till the data was loaded (which will take time)

### Code example
I was able to narrow down the problem to the following location in the [code](https://github.com/pimcore/pimcore/blob/master/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php#L141):

```php
# pimcore/bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php:141
if (!$this->getAdminUser()->isAdmin()) {
    $userIds = $this->getAdminUser()->getRoles();
    $userIds[] = $this->getAdminUser()->getId();
    $condition .= ' AND (
                    (SELECT list FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') AND LOCATE(CONCAT(objects.o_path,objects.o_key),cpath)=1 ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
                    OR
                    (SELECT list FROM users_workspaces_object WHERE userId IN (' . implode(',', $userIds) . ') AND LOCATE(cpath,CONCAT(objects.o_path,objects.o_key))=1 ORDER BY LENGTH(cpath) DESC LIMIT 1)=1
                )';
}
```

The subqueries within the condition will take most of the time because these are executed for each element in the objects table. 

stats of a real-world EXPLAIN-statement (with obscured ids for the parent and user (roles)))
```
EXPLAIN SELECT
  objects.o_id as o_id,
  objects.o_type
FROM
  objects
WHERE
  (
    objects.o_parentId = 1
    AND (
      (
        SELECT
          list
        FROM
          users_workspaces_object
        WHERE
          userId IN (1,2,3)
          AND LOCATE(CONCAT(objects.o_path, objects.o_key), cpath) = 1
        ORDER BY
          LENGTH(cpath) DESC
        LIMIT
          1
      ) = 1
      OR (
        SELECT
          list
        FROM
          users_workspaces_object
        WHERE
          userId IN (1,2,3)
          AND LOCATE(cpath, CONCAT(objects.o_path, objects.o_key)) = 1
        ORDER BY
          LENGTH(cpath) DESC
        LIMIT
          1
      ) = 1
    )
    AND objects.o_type IN ('object', 'folder')
  )
ORDER BY
  CAST(objects.o_key AS CHAR CHARACTER SET utf8) COLLATE utf8_general_ci ASC
LIMIT
30
```

| id | select_type | table | type | possible_keys | key | key_len | ref | rows | Extra |
|:--|:--|:--|:--|:--|:--|:--|:--|:--|:--|
| 1 | PRIMARY | objects | ref | parentId,type | parentId | 5 | const | 584857 | Using where; Using filesort |
| 2 | DEPENDENT SUBQUERY | users_workspaces_object | range | userId | userId | 4 | NULL | 19 | Using index condition; Using where; Using filesort |
| 3 | DEPENDENT SUBQUERY | users_workspaces_object | range | userId | userId | 4 | NULL | 19 | Using index condition; Using where; Using filesort |

### PR DESCRIPTION
**Pull request by kingjia90**
improves #6383

### CODE REVIEW

**Comment by kingjia90** [comment]
getUserPermissions required 14 consecutive queries (one for each permission/column), this PR changes reduce to 3.
Due the nature of the user level permission having more power than role level permission, i am trying to "shortcircuit" when found one by `if ($parentRow['userId'] == $currentUserId){`, 
if not found we check the roles (having the same `cid` from first query, it guarantees that there are role level but not any workspace at the user level) and do the additive calculation (breaking the sub-loop as soon as permission=1 is found).
Another potential benefit of this approach is that when `cid` for the element itself is not found, we check the parent one on the second query and the db engine should be internally "cache" it, compared to the [isAllowed](https://github.com/pimcore/pimcore/blob/89ecd1b81d4e955164027ca3a9129fa4b0d82679/models/DataObject/AbstractObject/Dao.php#L499) which is more one-time use only.

some reference values that i have noticed during manual tests:
when listing a folder with 30 data objects (the default max per pagination)

admin: 200-250ms
with this PR: 450-500ms
with #11714 changes: 600-750ms
prior #11714: 900ms-1s+

more details here: https://github.com/pimcore/pimcore/issues/6383#issuecomment-1073309284

**Comment by kingjia90** [comment]
Both tests  (before/after the PR) are matching and passing.


Changes required to test the before:
In pimcore\models\Element\AbstractElement.php
add `$user` parameter on
`public function getUserPermissions($user = null)` and `$permissions[$name] = $this->isAllowed($name, $user);`
then `public function isAllowed` should return (int) instead of (bool), because `false` would just be empty and in the test i am using `0`, so the array_diff_assoc wrongly triggers.

**Review by fashxp — CHANGES_REQUESTED** [review]
- [x] add note to docs that for list permission in listings, only further restricting, but not further allowing is possible. 
list=1 -> list=0 --> possible
list=0 -> list=1 --> NOT possible

File: tests/Model/Permissions/ModelDocumentPermissionsTest.php [hunk 1]
```diff
@@ -301,6 +309,52 @@ public function testIsAllowed()
...
+        // User 1 doesn't have publish=1 based on the inherited `Testrole`
+        $this->doAreAllowedTest($this->groupfolder, ['save'=> 1,'publish' =>1], true, false, true);
+        // Check admin and User2 are not simply returning all true to pass the above test, confirms User2's user-level permissions getting priority over individual roles
+        $this->doAreAllowedTest($this->groupfolder, ['save'=> 1,'publish' =>1 ,'versions' => 0], false, false, true);
+        // `Dummyrole`(0,0) is the "weakest" when conflicting, so it should never true, User 2 doesn't have user-level permission, so this confirms the additive behaviour on conflicting roles
+        $this->doAreAllowedTest($this->groupfolder, ['save'=> 0,'publish' =>0], false, false, false);
+        // Checks User1 is correctly getting the workspace rule from roles
+        $this->doAreAllowedTest($this->groupfolder, ['save'=> 1,'publish' =>0], false, true, false);
+
+//        //check when no parent workspace is found, it should be allow list=1 when children are found, in this case for admin and user1 to get to `c`
```
Comment (fashxp) [inline tests/Model/Permissions/ModelDocumentPermissionsTest.php]: why this test is commented?

File: tests/Model/Permissions/ModelAssetPermissionsTest.php [hunk 2]
```diff
@@ -301,6 +309,52 @@ public function testIsAllowed()
...
+        // User 1 doesn't have publish=1 based on the inherited `Testrole`
+        $this->doAreAllowedTest($this->groupfolder, ['delete'=> 1,'publish' =>1], true, false, true);
+        // Check admin and User2 are not simply returning all true to pass the above test, confirms User2's user-level permissions getting priority over individual roles
+        $this->doAreAllowedTest($this->groupfolder, ['delete'=> 1,'publish' =>1 ,'versions' => 0], false, false, true);
+        // `Dummyrole`(0,0) is the "weakest" when conflicting, so it should never true, User 2 doesn't have user-level permission, so this confirms the additive behaviour on conflicting roles
+        $this->doAreAllowedTest($this->groupfolder, ['delete'=> 0,'publish' =>0], false, false, false);
+        // Checks User1 is correctly getting the workspace rule from roles
+        $this->doAreAllowedTest($this->groupfolder, ['delete'=> 1,'publish' =>0], false, true, false);
+
+        //check when no parent workspace is found, it should be allow list=1 when children are found, in this case for admin and user1 to get to `c`
```
Comment (fashxp) [inline tests/Model/Permissions/ModelAssetPermissionsTest.php]: why this test is commented?

**Review by fashxp — CHANGES_REQUESTED** [review]

**Comment by fashxp** [comment]
- [x] I think we should rebase that for 10.x ... as it is no bugfix, but an improvement.

**Comment by fashxp** [comment]
we first merge https://github.com/pimcore/pimcore/pull/11938 ... and then we can merge that one

**Review by fashxp — APPROVED** [review]

**Comment by fashxp** [comment]
my tests show following improvements with a folder of 20.000 direct child objects on a default local docker setup on a windows machine:  

| action | 10.3.4 | prior that PR | with that PR | 
| ------- | ------ | -------------- | ------------- |
| grid proxy (loading 25 items) | 1.3 sec | 1.3 sec | 1.0 sec |
| tree get childs (loading 30 items) | 2.7sec | 1.5 sec | 1.2 sec |

I did not notice any significant changes for admin users, they always were around 300ms per request. 

That's what we can archive right now.

**Comment by kingjia90** [comment]
Interesting data, It would be nice to have the tests for grid proxy with pagination 200 (how that 300ms improvement is by scaling up) and one for one item in one folder (to see the minimum required time no matter how many items)

**Pull request merged**